src/odemis/acq/fastem.py

```diff
@@ -18,73 +20,398 @@
+            self._future.set_progress(end=time.time() + expected_time)
+
+        return self.megafield
+
+    def image_received(self, dataflow, data):
+        """
+        Function called by dataflow when data has been received from the detector.
+        :param dataflow: (model.DataFlow) The dataflow on the detector.
+        :param data: (model.DataArray) The data array containing the image data.
+        """
+        self.megafield[self.field_idx] = data
+        # When data is received notify the threading event, which keeps track of whether data was received.
+        self._data_received.set()
+
+    def cancel(self, future):
+        """
+        Cancels the ROA acquisition.
+        :param future: (future) The ROA (megafield) future.
+        :return: (bool) True if cancelled, False if too late to cancel as future is already finished.
+        """
+        self._cancelled = True
+
+        # Report if it's too late for cancellation (and the f.result() will return)
+        if not self._fields_remaining:
+            return False
+
+        return True
+
+    def get_abs_stage_movement(self):
+        """
+        Based on the field index calculate the stage position where the next tile (field image) should be acquired.
+        The position is always calculated with respect to the top/left tile (field image). The stage position returned
+        is the center of the respective tile.
+        :return: (float, float) The new absolute stage x and y position in meter.
+        """
+        px_size = self._multibeam.pixelSize.value
+        field_res = self._multibeam.resolution.value
+        pos_orig = self._roa.coordinates.value[:2]  # position of top/left corner of the ROA
+
+        # The position of the stage when acquiring the top/left tile needs to be matching the center of the tile.
+        pos_first_tile = (pos_orig[0] - field_res[0]/2. * px_size[0], pos_orig[1] + field_res[1]/2. * px_size[1])
+
+        rel_move_hor = self.field_idx[0] * px_size[0] * field_res[0]  # in meter
+        rel_move_vert = self.field_idx[1] * px_size[1] * field_res[1]  # in meter
+
+        # Move in the negative x direction, because the second field should be right of the first.
+        # Move in positive y direction, because the second field should be bottom of the first.
+        # TODO verify that this results in the images being taken in the correct order (left to right, top to bottom).
+        pos_hor = pos_first_tile[0] - rel_move_hor
+        pos_vert = pos_first_tile[1] + rel_move_vert
+
+        return pos_hor, pos_vert
 
+    def move_stage_to_next_tile(self):
+        """Move the stage to the next tile (field image) position."""
 
-def estimateTime(roas):
-    return len(roas) * 2
+        pos_hor, pos_vert = self.get_abs_stage_movement()  # get the absolute position for the new tile
+        f = self._stage.moveAbs({'x': pos_hor, 'y': pos_vert})  # move the stage
+        f.result()
```

Comment:
Yes good idea. I implemented it like this:
`        try:
            f.result(timeout=100)  # don't wait forever
            logging.debug("Moved to stage position %s" % (self._stage.position.value,))
        except TimeoutError:
            self._cancelled = True  # cancel the acquisition
            logging.warning("Stage movement failed.")`
            
Then the `self._cancelle`d flag should be detected in the `acquire_ro`a (but only after executing another field scan, which is not ideal) and thus return to the `ru`n method, where the cancellation error is raised. Is it better to directly cancel the `self._future` here instead?

Reply:
I don't think it should cancel the acquisition. "Cancel" is in case the user changes his mind. If there is a problem with the hardware, you want to raise a "proper" error. A TimeoutError (explaining that the stage failed to move after 100s when requesting to go to pos P) would be a more appropriate exception. Then it should trickled down until run() and stop the whole acquisition.